pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+
+An example documenting the attributes of a class, or in this case, the keys
+of a ``TypedDict``, could look like this:
+
+.. code-block::
+
+    from typing import Annotated, TypedDict, NotRequired, doc
+
+
+    class User(TypedDict):
+        firstname: Annotated[str, doc("The user's first name")]
+        lastname: Annotated[str, doc("The user's last name")]
+
+
+An example documenting the parameters of a function could look like this:
+
+.. code-block::
+
+    from typing import Annotated, doc
+
+
+    def create_user(
+        lastname: Annotated[str, doc("The **last name** of the newly created user")],
+        firstname: Annotated[str | None, doc("The user's **first name**")] = None,
+    ) -> Annotated[User, doc("The created user after saving in the database")]:
+        """
+        Create a new user in the system, it needs the database connection to be already
+        initialized.
+        """
+        pass
+
+
+The return of the ``doc()`` function is an instance of a class that can be checked
+and used at runtime, defined similar to:
+
+.. code-block::
+
+    class DocInfo:
+        def __init__(self, documentation: str):
+            self.documentation = documentation
+
+...where the attribute ``documentation`` contains the same value string passed to
+the function ``doc()``.
+
+
+Additional Scenarios
+--------------------
+
+The main scenarios that this proposal intends to cover are described above, and
+for implementers to be conformant to this specification, they only need to support
+those scenarios described above.
+
+Here are some additional edge case scenarios with their respective considerations,
+but implementers are not required to support them.
+
+
+Type Alias
+----------
+
+When creating a type alias, like:
+
+.. code-block::
+
+    Username = Annotated[str, doc("The name of a user in the system")]
+
+
+...the documentation would be considered to be carried by the parameter annotated
+with ``Username``.
+
+So, in a function like:
+
+.. code-block::
+
+    def hi(
+      to: Username,
+    ) -> None: ...
+
+
+...it would be equivalent to:
+
+.. code-block::
+
+    def hi(
+      to: Annotated[str, doc("The name of a user in the system")],
+    ) -> None: ...
+
+Nevertheless, implementers would not be required to support type aliases outside
+of the final type annotation to be conformant with this specification, as it
+could require more complex dereferencing logic.
+
+
+Annotating Type Parameters
+--------------------------
+
+When annotating type parameters, as in:
+
+.. code-block::
+
+    def hi(
+      to: list[Annotated[str, doc("The name of a user in a list")]],
+    ) -> None: ...
+
+...the documentation in ``doc()`` would refer to what it is annotating, in this
+case, each item in the list, not the list itself.
+
+There are currently no practical use cases for documenting type parameters,
+so implementers are not required to support this scenario to be considered
+conformant, but it's included for completeness.
```

Comment:
I'm unsure what this section means -- what do you mean by a type parameter or there being no usecases?